File: android/src/main/java/com/volla/expo/ivs/player/ExpoIvsPlayerView.kt
package com.volla.expo.ivs.player

import android.content.Context
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.view.ViewGroup
import com.amazonaws.ivs.player.*
import expo.modules.kotlin.AppContext
import expo.modules.kotlin.viewevent.EventDispatcher
import expo.modules.kotlin.views.ExpoView
import java.util.*

class ExpoIvsPlayerView(context: Context, appContext: AppContext) : ExpoView(context, appContext) {
  
  // Event dispatchers
  private val onSeek by EventDispatcher()
  private val onData by EventDispatcher()
  private val onVideoStatistics by EventDispatcher()
  private val onPlayerStateChange by EventDispatcher()
  private val onDurationChange by EventDispatcher()
  private val onQualityChange by EventDispatcher()
  private val onPipChange by EventDispatcher()
  private val onRebuffering by EventDispatcher()
  private val onLoadStart by EventDispatcher()
  private val onLoad by EventDispatcher()
  private val onLiveLatencyChange by EventDispatcher()
  private val onTextCue by EventDispatcher()
  private val onTextMetadataCue by EventDispatcher()
  private val onProgress by EventDispatcher()
  private val onError by EventDispatcher()
  private val onTimePoint by EventDispatcher()
  
  // IVS Player components
  private var playerView: PlayerView? = null
  private var player: Player? = null
  private var progressTimer: Timer? = null
  private val sources = mutableMapOf<Int, String>()
  private val handler = Handler(Looper.getMainLooper())
  
  // Properties
  var streamUrl: String? = null
    set(value) {
      field = value
      value?.let { loadStream(it) }
    }
  
  var paused: Boolean = false
    set(value) {
      field = value
      updatePlayPauseState()
    }
  
  var muted: Boolean = false
    set(value) {
      field = value
      player?.isMuted = value
    }
  
  var volume: Float = 1.0f
    set(value) {
      field = value
      player?.volume = value
    }
  
  var playbackRate: Float = 1.0f
    set(value) {
      field = value
      player?.playbackRate = value
    }
  
  var autoplay: Boolean = false
  var loop: Boolean = false
  var liveLowLatency: Boolean = true
  var rebufferToLive: Boolean = false
  var resizeMode: String = "aspectFit"
    set(value) {
      field = value
      setupVideoScaling()
    }
  
  var progressInterval: Double = 1.0
  var pipEnabled: Boolean = false
  var logLevel: String = "ERROR"
  var maxBitrate: Int = 0
  var initialBufferDuration: Double = 0.0
  var autoQualityMode: Boolean = true
  
  init {
    setupPlayer()
  }
  
  private fun setupPlayer() {
    try {
      // Create player view
      playerView = PlayerView(context)
      playerView?.layoutParams = ViewGroup.LayoutParams(
        ViewGroup.LayoutParams.MATCH_PARENT,
        ViewGroup.LayoutParams.MATCH_PARENT
      )
      
      // Create player
      player = Player.create(context)
      playerView?.player = player
      
      // Add player view to this view
      playerView?.let { addView(it) }
      
      // Setup player listeners
      setupPlayerListeners()
      setupVideoScaling()
      
    } catch (e: Exception) {
      Log.e("ExpoIvsPlayerView", "Error setting up player: ${e.message}")
      onError(mapOf("error" to "Player setup failed: ${e.message}"))
    }
  }
  
  private fun setupPlayerListeners() {
    player?.let { player ->
      
      // Player state listener
      player.addListener(object : Player.Listener {
        override fun onStateChanged(state: Player.State) {
          val stateString = when (state) {
            Player.State.IDLE -> {
              stopProgressTimer()
              "Idle"
            }
            Player.State.READY -> {
              onLoad(mapOf("duration" to player.duration))
              
              // Send data event with qualities and session info
              val qualities = player.qualities.map { quality ->
                mapOf(
                  "name" to quality.name,
                  "codecs" to quality.codecs,
                  "bitrate" to quality.bitrate,
                  "framerate" to (quality.framerate ?: 0f),
                  "width" to (quality.width ?: 0),
                  "height" to (quality.height ?: 0)
                )
              }
              
              onData(mapOf(
                "qualities" to qualities,
                "version" to player.version,
                "sessionId" to (player.sessionId ?: "")
              ))
              
              if (autoplay && !paused) {
                player.play()
              }
              "Ready"
            }
            Player.State.BUFFERING -> {
              onRebuffering(emptyMap())
              "Buffering"
            }
            Player.State.PLAYING -> {
              startProgressTimer()
              "Playing"
            }
            Player.State.ENDED -> {
              stopProgressTimer()
              if (loop) {
                player.seekTo(0L)
                player.play()
              }
              "Ended"
            }
          }
          
          onPlayerStateChange(mapOf("state" to stateString))
        }
        
        override fun onDurationChanged(duration: Long) {
          val durationSeconds = if (duration >= 0) duration / 1000.0 else null
          onDurationChange(mapOf("duration" to durationSeconds))
        }
        
        override fun onQualityChanged(quality: Quality?) {
          if (quality != null) {
            onQualityChange(mapOf(
              "name" to quality.name,
              "codecs" to quality.codecs,
              "bitrate" to quality.bitrate,
              "framerate" to (quality.framerate ?: 0f),
              "width" to (quality.width ?: 0),
              "height" to (quality.height ?: 0)
            ))
          } else {
            onQualityChange(mapOf("quality" to null))
          }
        }
        
        override fun onError(exception: PlayerException) {
          this@ExpoIvsPlayerView.onError(mapOf("error" to exception.errorMessage))
        }
        
        override fun onRebuffering() {
          onRebuffering(emptyMap())
        }
        
        override fun onSeekCompleted(position: Long) {
          onSeek(mapOf("position" to position / 1000.0))
        }
        
        override fun onVideoStatistics(statistics: VideoStatistics) {
          onVideoStatistics(mapOf(
            "bitrate" to statistics.bitrate,
            "duration" to (statistics.duration ?: 0L) / 1000.0,
            "framesDecoded" to (statistics.framesDecoded ?: 0L),
            "framesDropped" to (statistics.framesDropped ?: 0L)
          ))
        }
        
        override fun onCue(cue: Cue) {
          when (cue) {
            is TextCue -> {
              onTextCue(mapOf(
                "type" to "text",
                "text" to cue.text,
                "line" to cue.line,
                "size" to cue.size,
                "position" to cue.position,
                "textAlignment" to cue.textAlignment.name
              ))
            }
            is TextMetadataCue -> {
              onTextMetadataCue(mapOf(
                "type" to "metadata",
                "text" to (cue.text ?: ""),
                "textDescription" to (cue.textDescription ?: "")
              ))
            }
          }
        }
        
        override fun onMetadata(type: String, data: String) {
          // Handle metadata if needed
        }
        
        override fun onLiveLatencyChanged(latency: Long) {
          onLiveLatencyChange(mapOf("latency" to latency))
        }
      })
    }
  }
  
  private fun setupVideoScaling() {
    playerView?.let { view ->
      when (resizeMode) {
        "aspectFill" -> view.videoScaleType = PlayerView.VideoScaleType.ZOOM
        "aspectFit" -> view.videoScaleType = PlayerView.VideoScaleType.FIT
        "aspectZoom" -> view.videoScaleType = PlayerView.VideoScaleType.FILL
        else -> view.videoScaleType = PlayerView.VideoScaleType.FIT
      }
    }
  }
  
  private fun loadStream(urlString: String) {
    try {
      onLoadStart(emptyMap())
      val uri = android.net.Uri.parse(urlString)
      player?.load(uri)
    } catch (e: Exception) {
      onError(mapOf("error" to "Invalid URL: $urlString"))
    }
  }
  
  private fun updatePlayPauseState() {
    if (paused) {
      player?.pause()
    } else {
      player?.play()
    }
  }
  
  private fun startProgressTimer() {
    stopProgressTimer()
    progressTimer = Timer()
    progressTimer?.scheduleAtFixedRate(object : TimerTask() {
      override fun run() {
        player?.let { player ->
          handler.post {
            val position = player.position / 1000.0
            onProgress(mapOf("progress" to position))
          }
        }
      }
    }, 0, (progressInterval * 1000).toLong())
  }
  
  private fun stopProgressTimer() {
    progressTimer?.cancel()
    progressTimer = null
  }
  
  // MARK: - Commands
  
  fun preload(url: String, sourceId: Int) {
    sources[sourceId] = url
  }
  
  fun loadSource(sourceId: Int) {
    sources[sourceId]?.let { url ->
      loadStream(url)
    }
  }
  
  fun releaseSource(sourceId: Int) {
    sources.remove(sourceId)
  }
  
  fun play() {
    player?.play()
  }
  
  fun pause() {
    player?.pause()
  }
  
  fun seekTo(position: Double) {
    val positionMs = (position * 1000).toLong()
    player?.seekTo(positionMs)
  }
  
  fun setOrigin(origin: String) {
    // Set origin for network requests if needed
    // This might require custom implementation based on requirements
  }
  
  fun togglePip() {
    // Picture-in-picture support would require additional Android PiP implementation
    // For now, dispatch event to indicate state
    onPipChange(mapOf("isActive" to false))
  }
  
  override fun onDetachedFromWindow() {
    super.onDetachedFromWindow()
    stopProgressTimer()
    player?.release()
  }
}

File: ios/ExpoIvsPlayerView.swift
import ExpoModulesCore
import AmazonIVSPlayer
import UIKit
import AVKit

class ExpoIvsPlayerView: ExpoView, IVSPlayer.Delegate {
  private var playerView: IVSPlayerView?
  private var player: IVSPlayer?
  private var pipController: AVPictureInPictureController?
  private var progressTimer: Timer?
  private var sources: [Int: String] = [:]
  
  // Event handlers
  let onSeek = EventDispatcher()
  let onData = EventDispatcher()
  let onVideoStatistics = EventDispatcher()
  let onPlayerStateChange = EventDispatcher()
  let onDurationChange = EventDispatcher()
  let onQualityChange = EventDispatcher()
  let onPipChange = EventDispatcher()
  let onRebuffering = EventDispatcher()
  let onLoadStart = EventDispatcher()
  let onLoad = EventDispatcher()
  let onLiveLatencyChange = EventDispatcher()
  let onTextCue = EventDispatcher()
  let onTextMetadataCue = EventDispatcher()
  let onProgress = EventDispatcher()
  let onError = EventDispatcher()
  let onTimePoint = EventDispatcher()
  
  // Properties
  var streamUrl: String? {
    didSet {
      if let url = streamUrl {
        loadStream(url)
      }
    }
  }
  
  var paused: Bool = false {
    didSet {
      updatePlayPauseState()
    }
  }
  
  var muted: Bool = false {
    didSet {
      player?.muted = muted
    }
  }
  
  var volume: Float = 1.0 {
    didSet {
      playerView?.player?.volume = volume
    }
  }
  
  var playbackRate: Float = 1.0 {
    didSet {
      player?.playbackRate = playbackRate
    }
  }
  
  var autoplay: Bool = false
  var loop: Bool = false
  var liveLowLatency: Bool = true
  var rebufferToLive: Bool = false
  var resizeMode: String = "aspectFit"
  var progressInterval: Double = 1.0
  var pipEnabled: Bool = false
  var logLevel: String = "IVSLogLevelError"
  var maxBitrate: Int = 0
  var initialBufferDuration: Double = 0
  var autoQualityMode: Bool = true
  var showControls: Bool = false {
    didSet {
      // IVSPlayerView doesn't have built-in controls on iOS
      // This prop is added for API consistency with Android
      // Custom controls can be implemented if needed
    }
  }
  
  required init(appContext: AppContext? = nil) {
    super.init(appContext: appContext)
    setupPlayer()
  }
  
  private func setupPlayer() {
    playerView = IVSPlayerView()
    player = IVSPlayer()
    player?.delegate = self
    playerView?.player = player
    
    if let playerView = playerView {
      addSubview(playerView)
      playerView.translatesAutoresizingMaskIntoConstraints = false
      NSLayoutConstraint.activate([
        playerView.topAnchor.constraint(equalTo: topAnchor),
        playerView.leadingAnchor.constraint(equalTo: leadingAnchor),
        playerView.trailingAnchor.constraint(equalTo: trailingAnchor),
        playerView.bottomAnchor.constraint(equalTo: bottomAnchor)
      ])
    }
    
    setupVideoGravity()
    setupPictureInPicture()
  }
  
  func setupVideoGravity() {
    guard let playerView = playerView else { return }
    
    switch resizeMode {
    case "aspectFill":
      playerView.videoGravity = .resizeAspectFill
    case "aspectFit":
      playerView.videoGravity = .resizeAspect
    case "aspectZoom":
      playerView.videoGravity = .resize
    default:
      playerView.videoGravity = .resizeAspect
    }
  }
  
  func setupPictureInPicture() {
    guard pipEnabled else { return }
    
    // PiP is not directly supported with IVSPlayerView
    // This would require custom implementation with AVPlayerLayer
    // For now, we'll skip PiP support for IVS Player
  }
  
  private func loadStream(_ urlString: String) {
    guard let url = URL(string: urlString) else {
      onError(["error": "Invalid URL: \(urlString)"])
      return
    }
    
    onLoadStart([:])
    player?.load(url)
  }
  
  private func updatePlayPauseState() {
    if paused {
      player?.pause()
    } else {
      player?.play()
    }
  }
  
  private func startProgressTimer() {
    stopProgressTimer()
    progressTimer = Timer.scheduledTimer(withTimeInterval: progressInterval, repeats: true) { [weak self] _ in
      guard let self = self, let player = self.player else { return }
      let position = player.position
      self.onProgress(["progress": position])
    }
  }
  
  private func stopProgressTimer() {
    progressTimer?.invalidate()
    progressTimer = nil
  }
  
  // MARK: - Commands
  
  func preload(url: String, sourceId: Int) {
    sources[sourceId] = url
  }
  
  func loadSource(sourceId: Int) {
    if let url = sources[sourceId] {
      loadStream(url)
    }
  }
  
  func releaseSource(sourceId: Int) {
    sources.removeValue(forKey: sourceId)
  }
  
  func play() {
    player?.play()
  }
  
  func pause() {
    player?.pause()
  }
  
  func seekTo(position: Double) {
    player?.seek(to: CMTime(seconds: position, preferredTimescale: 1000)) { [weak self] _ in
      self?.onSeek(["position": position])
    }
  }
  
  func setOrigin(origin: String) {
    // Set origin for network requests if needed
  }
  
  func togglePip() {
    if pipController?.isPictureInPictureActive == true {
      pipController?.stopPictureInPicture()
    } else {
      pipController?.startPictureInPicture()
    }
  }
  
  // MARK: - IVSPlayerDelegate
  
  func player(_ player: IVSPlayer, didChangeState state: IVSPlayer.State) {
    let stateString: String
    switch state {
    case .idle:
      stateString = "Idle"
      stopProgressTimer()
    case .ready:
      stateString = "Ready"
      onLoad(["duration": player.duration])
      
      let qualities = player.qualities
      if !qualities.isEmpty {
        let qualitiesData = qualities.map { quality in
          return [
            "name": quality.name,
            "codecs": quality.codecs,
            "bitrate": quality.bitrate,
            "framerate": quality.framerate,
            "width": quality.width,
            "height": quality.height
          ]
        }
        
        onData([
          "qualities": qualitiesData,
          "version": player.version,
          "sessionId": player.sessionId
        ])
      }
      
      if autoplay && !paused {
        player.play()
      }
    case .buffering:
      stateString = "Buffering"
      onRebuffering([:])
    case .playing:
      stateString = "Playing"
      startProgressTimer()
    case .ended:
      stateString = "Ended"
      stopProgressTimer()
      if loop {
        player.seek(to: .zero) { _ in
          player.play()
        }
      }
    @unknown default:
      stateString = "Unknown"
    }
    
    onPlayerStateChange(["state": stateString])
  }
  
  func player(_ player: IVSPlayer, didOutputCue cue: IVSCue) {
    switch cue {
    case let textCue as IVSTextCue:
      onTextCue([
        "type": "text",
        "text": textCue.text,
        "line": textCue.line,
        "size": textCue.size,
        "position": textCue.position,
        "textAlignment": String(describing: textCue.textAlignment)
      ])
    case let metadataCue as IVSTextMetadataCue:
      onTextMetadataCue([
        "type": "metadata",
        "text": metadataCue.text,
        "textDescription": metadataCue.textDescription
      ])
    default:
      break
    }
  }
  
  func player(_ player: IVSPlayer, didChangeQuality quality: IVSQuality?) {
    if let quality = quality {
      onQualityChange([
        "name": quality.name,
        "codecs": quality.codecs,
        "bitrate": quality.bitrate,
        "framerate": quality.framerate,
        "width": quality.width,
        "height": quality.height
      ])
    } else {
      onQualityChange([:])
    }
  }
  
  func player(_ player: IVSPlayer, didChangeDuration duration: CMTime) {
    let durationSeconds = duration.isValid && !duration.isIndefinite ? duration.seconds : nil
    onDurationChange(["duration": durationSeconds as Any])
  }
  
  func player(_ player: IVSPlayer, didFailWithError error: Error) {
    onError(["error": error.localizedDescription])
  }
  
  func playerWillRebuffer(_ player: IVSPlayer) {
    onRebuffering([:])
  }
  
  deinit {
    stopProgressTimer()
    player?.delegate = nil
  }
}

// MARK: - AVPictureInPictureControllerDelegate

extension ExpoIvsPlayerView: AVPictureInPictureControllerDelegate {
  func pictureInPictureControllerWillStartPictureInPicture(_ pictureInPictureController: AVPictureInPictureController) {
    onPipChange(["isActive": true])
  }
  
  func pictureInPictureControllerDidStopPictureInPicture(_ pictureInPictureController: AVPictureInPictureController) {
    onPipChange(["isActive": false])
  }
}
